src/azul/drs.py

```diff
@@ -74,10 +87,139 @@ class Access:
+        subcls = CompactDRSURI if drs_uri.find(':', len(scheme)) >= 0 else RegularDRSURI
+        return subcls.parse(drs_uri)
+
+    @abstractmethod
+    def to_url(self, client: 'DRSClient', access_id: Optional[str] = None) -> str:
+        """
+        Translate the DRS URI into a DRS URL. All query params included in the DRS
+        URI (eg '{drs_uri}?version=123') will be carried over to the DRS URL.
+        """
+        raise NotImplementedError
+
+
+@attr.s(auto_attribs=True, kw_only=True, frozen=True, slots=True)
+class RegularDRSURI(DRSURI):
+    uri: furl
+
+    def __attrs_post_init__(self):
+        assert self.uri.scheme == 'drs', self.uri
+
+    @property
+    def netloc(self) -> str:
+        return self.uri.netloc
+
+    @classmethod
+    def parse(cls, drs_uri: str) -> 'RegularDRSURI':
+        return cls(uri=furl(drs_uri))
+
+    def to_url(self, client: 'DRSClient', access_id: Optional[str] = None) -> str:
+        url = self.uri.copy().set(scheme='https')
+        url.set(path=drs_object_url_path(object_id=one(self.uri.path.segments),
+                                         access_id=access_id))
+        return str(url)
+
+
+@attr.s(auto_attribs=True, kw_only=True, frozen=True, slots=True)
+class CompactDRSURI(DRSURI):
+    """
+    So-called DRS "URIs" [1] for Compact Identifiers [2] are NOT URIs according
+    to RFC 3986 [3] so we can't use off-the-shelf URI parsers.
+
+    [1] https://ga4gh.github.io/data-repository-service-schemas/preview/release/drs-1.3.0/docs/
+
+    [2] https://www.nature.com/articles/sdata201829
+
+    [3] https://datatracker.ietf.org/doc/html/rfc3986
+    """
+    _namespace: str
+    _accession: str
+
+    def __attrs_post_init__(self):
+        assert '/' not in self._namespace and '?' not in self._accession, self
+
+    @classmethod
+    def parse(cls, drs_uri: str) -> 'CompactDRSURI':
+        scheme, netloc = drs_uri.split('://', 1)
+        # Compact identifier-based URIs can be hard to parse when following
+        # RFC3986, with the 'namespace:accession' part matching either the
+        # heir-part or path production depending if the optional provider code
+        # and following slash is included.
+        # https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#compact-identifier-based-drs-uris
+        prefix, accession = netloc.split(':', 1)
+        reject('/' in prefix,
+               'Compact identifiers with provider codes are not supported', drs_uri)
+        reject('?' in accession,
+               'Compact identifiers must not contain query parameters', drs_uri)
+        return cls(namespace=prefix,
+                   accession=accession)
+
+    def to_url(self, client: 'DRSClient', access_id: Optional[str] = None) -> str:
+        url = client.id_client.resolve(self._namespace, self._accession)
+        require(str(url.path) == drs_object_url_path(object_id=self._accession),
```

Comment:
Add brief comment explaining this. The key observation is that there are two ways to arrive at the same path, one by following the DRS spec, the other by expanding the URL pattern registered at identifiers.org for the namespace, except that the latter can't be used to embed the access ID.